tests/js/common/shell/shell-keygen.js

```diff
@@ -32,29 +32,246 @@ const jsunity = require("jsunity");
+
+    // 13 documents inserted so far. Try to insert some malformed ones (must fail).
+    insertInvalidSmartVertices(vn);
+
+    assertEqual(13, db[vn].count());
+  } finally {
+    graphs._drop(gn, true);
+  }
+};
+
+const runSmartEdgeInserts = (graphProperties) => {
+  // Helper method to either insert edges into SmartGraphs or EnterpriseGraphs
+  if (cluster) {
+    // fail if we generate a key on a DB server
+    debugSetFailAtAll("KeyGenerator::generateOnSingleServer");
+  } else {
+    // single server: we can actually get here with the SmartGraph simulator!
+    debugSetFailAtAll("KeyGenerator::generateOnCoordinator");
+  }
+
+  graphs._create(gn, [graphs._relation(en, vn, vn)], null, graphProperties);
+
+  try {
+    // test various ways of inserting documents into the collection
+
+    // single insert, using document API
+    db[en].insert({value: "42", _from: vn + "/test:42", _to: vn + "/test:42"});
+
+    // batch insert, using document API
+    db[en].insert([{value: "42", _from: vn + "/test:42", _to: vn + "/test:42"}, {
+      value: "42",
+      _from: vn + "/test:42",
+      _to: vn + "/test:42"
+    }, {value: "42", _from: vn + "/test:42", _to: vn + "/test:42"}]).forEach((res) => {
+      assertFalse(res.hasOwnProperty('error'));
+    });
+
+    // single insert, using AQL
+    db._query(`INSERT
+          { value: "42", _from: "${vn}/test:42", _to: "${vn}/test:42" } INTO
+          ${en}`);
+    db._query(`INSERT
+          { value: "42", _from: "${vn}/test:42", _to: "${vn}/test:42" } INTO
+          ${en}`, null, disableSingleDocRule);
+    db._query(`INSERT
+          { value: "42", _from: "${vn}/test:42", _to: "${vn}/test:42" } INTO
+          ${en}`, null, disableRestrictToSingleShardRule);
+
+    // batch insert, using AQL
+    db._query(`FOR i IN 1..3 INSERT { value: "42", _from: "${vn}/test:42", _to: "${vn}/test:42" } INTO ${en}`);
+    db._query(`FOR i IN 1..3 INSERT { value: "42", _from: "${vn}/test:42", _to: "${vn}/test:42" } INTO ${en}`, null, disableRestrictToSingleShardRule);
+
+    // 13 documents inserted so far. Try to insert some malformed ones (must fail).
+    insertInvalidSmartEdges(en);
+
+    assertEqual(13, db[en].count());
+  } finally {
+    graphs._drop(gn, true);
+  }
+};
+
+const insertInvalidSmartVertices = (vn) => {
+  // after all positive tests, now some negative ones (not allowed inserts)
+  // we will not add any documents additionally!
+
+  const invalidKeys = [
+    "whitespace inbetween", // whitespace not allowed
+    12345, // numeric value not allowed
+    NaN,  // invalid numeric value not allowed
+    null // null not allowed
+  ];
+
+  // Note: Using save() <-> insert() should be the same.
+  // Using both here to double-check that the assumption is
+  // proven.
+
+  // single insert, using document API
+  invalidKeys.forEach(invalidKey => {
+    try {
+      db[vn].insert({_key: invalidKey, value: "43"});
+    } catch (ignore) {
```

Comment:
If this insert attempt is supposed to fail, we should add `fail()` after it for clarity.
I know at the end we are checking the number of docs in the collection, but checking for failures immediately here would be preferrable IMO.